_examples/hellocolibri/hellocolibri.go

```diff
@@ -0,0 +1,340 @@
+	res, err = daemon.ColibriSetupRsv(ctx, setupReq2)
+	check(err)
+	fmt.Println("admitted")
+	fmt.Printf("path type: %s\n", res.Path().Type)
+	colibriPath := res.Path()
+
+	//////////////////////////////////////////////////////////////////////////////////
+	//
+	// find the server address
+	pathquerier := sciond.Querier{
+		Connector: daemon,
+		IA:        localIA,
+	}
+	pathsToDst, err := pathquerier.Query(ctx, dstIA)
+	check(err)
+	p := pathsToDst[0]
+	fmt.Printf("Will connect to server using path %s\n", p)
+
+	serverUdpAddr, err := net.ResolveUDPAddr("udp", serverSciondPath)
+	check(err)
+	serverUdpAddr.Port = serverPort
+	serverAddr := &snet.UDPAddr{
+		IA:      dstIA,
+		Path:    p.Path(),
+		Host:    serverUdpAddr,
+		NextHop: p.UnderlayNextHop(),
+	}
+	// connect to the server
+	udpAddr, err := net.ResolveUDPAddr("udp", clientSciondPath)
+	check(err)
+	udpAddr.Port = 44321 // TODO(juagargi) or zero?
+
+	// connect to server
+	conn, err := scionNet.Dial(ctx, "udp", udpAddr, serverAddr, addr.SvcNone)
+	check(err)
+	_, err = conn.Write([]byte("hello there, best effort"))
+	check(err)
+	data := <-serverChan
+	fmt.Printf("we know the server got data: %s\n", string(data))
+	conn.Close()
+	//
+	//
+	// now dial using colibri
+	cp := &colpath.ColibriPath{}
+	err = cp.DecodeFromBytes(colibriPath.Raw)
+	check(err)
+	fmt.Printf("We will send traffic using colibri now, with path %s\n", printPath(cp))
+	serverAddr.Path.Type = colpath.PathType
+	serverAddr.Path.Raw = colibriPath.Raw
+
+	conn, err = scionNet.Dial(ctx, "udp", udpAddr, serverAddr, addr.SvcNone)
+	check(err)
+	_, err = conn.Write([]byte("hello there, colibri carried data"))
+	check(err)
+	data = <-serverChan
+	fmt.Printf("we know the server got data: %s\n", string(data))
+	conn.Close()
+
+	// clean the last reservation obtained
+	err = daemon.ColibriCleanupRsv(ctx, &setupReq2.Id, setupReq2.Index)
+	check(err)
+	fmt.Println("Reservation cleaned up")
+}
+
+func clientExtendedAPI(serverChan chan []byte) {
+	fmt.Println("client started")
+	ctx, cancelF := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancelF()
+
+	localUdpAddr, err := net.ResolveUDPAddr("udp", clientSciondPath)
+	check(err)
+	localUdpAddr.Port = 44321 // TODO(juagargi) or zero?
+	dstUdpAddr, err := net.ResolveUDPAddr("udp", serverSciondPath)
+	check(err)
+	dstUdpAddr.Port = serverPort
+	dstIA, err := addr.IAFromString("1-ff00:0:112")
+	check(err)
+
+	daemon, err := sciond.NewService(clientSciondPath).Connect(ctx)
+	check(err)
+	localIA, err := daemon.LocalIA(ctx)
+	check(err)
+	dispatcher := reliable.NewDispatcher(reliable.DefaultDispPath)
+	scionNet := snet.NewNetwork(localIA, dispatcher, sciond.RevHandler{Connector: daemon})
+	dstAddr := &snet.UDPAddr{
+		IA:   dstIA,
+		Host: dstUdpAddr,
+	}
+
+	fmt.Println("deleteme 1")
+	rsv, err := colapi.NewReservation(ctx, scionNet, daemon, dstAddr, 9, 0, func(a, b libcol.FullTrip) bool {
```

Comment:
When calling this twice in succession it would try to create two separate reservations. Would that work? Or should it return the same reservation for both calls?

Also might there be a use case for splitting this into two separate functions without going down to the level of the basic API? One function that simply returns a slice of `libcol.FullTrip`, and another function that selects one `libcol.FullTrip` of such a slice given one or more sorting predicates. This might be useful to query all possible `libcol.FullTrip`s in advance before making a selection later (I assume the querying, `daemon.ColibriListRsvs()`, takes some time).

When the reservation fails because the requested bandwidth is not supported it would make sense to be able to automatically fall back to the second-best path chosen via the given predicates and so on, until one supports the given bandwidth. In general, the available bandwidth should ideally be a parameter to base sorting on, just like the hop count and expiration time, such that trade-offs between the three of these can be defined rather than sorting based on only expiration time and hop count and hoping the requested bandwidth is supported. But it seems like the max bandwidth cannot be known in advance and can only be discovered by requesting an e2e reservation?

Reply:
> When calling this twice in succession it would try to create two separate reservations. Would that work? Or should it return the same reservation for both calls?

It would create two distinct reservations. It _should_ work :smiley: Note than creating a reservation is different from having an approved reservation, as the later happens when calling `Open`.

> Also might there be a use case for splitting this into two separate functions without going down to the level of the basic API? One function that simply returns a slice of `libcol.FullTrip`, and another function that selects one `libcol.FullTrip` of such a slice given one or more sorting predicates. This might be useful to query all possible `libcol.FullTrip`s in advance before making a selection later (I assume the querying, `daemon.ColibriListRsvs()`, takes some time).

Current API has it split in two parts: `New` and `Open`. The results of both are not visible to the caller, though. It might make sense to let the user "see" the result of the combination, as this is abstracted away from them through the sorting functions. Note that at the moment the API user can have the first sorting function always returning false just to keep track of the seen `FullTrip` in e.g. a map. Although this feels like a hack instead of a valid use of the API.

Reply:
> When the reservation fails because the requested bandwidth is not supported it would make sense to be able to automatically fall back to the second-best path chosen via the given predicates and so on, until one supports the given bandwidth. 

I would prevent this: when the reservation fails, the user can react quickly enough to emit a new reservation request, that will not take longer to be effective than using a "fallback" mechanism.
The scenario where the renewal fails is more time critical, to my eyes: when a renewal fails, the user has very limited time to get another reservation (less than 8 seconds), so an arbitrary number of failures cannot be solved that way, while it would benefit from having an in-protocol fallback mechanism (which we don't have), and this way the reservation renewal will be resolved in one go.

> In general, the available bandwidth should ideally be a parameter to base sorting on, just like the hop count and expiration time, such that trade-offs between the three of these can be defined rather than sorting based on only expiration time and hop count and hoping the requested bandwidth is supported. But it seems like the max bandwidth cannot be known in advance and can only be discovered by requesting an e2e reservation?

You are right, and I forgot to include the allocated bandwidth in the segment reservation look, i.e. `ReservationLooks` should contain much more information, like min/max/alloc bandwidth.
Note that the sorting functions do not have any restriction in what they use to sort, or if they want to keep state, etc. Any function has access to the `FullTrip`s they are sorting, which, as mentioned right above, will contain information such as allocated bandwidth.

Reply:
> I would prevent this: when the reservation fails, the user can react quickly enough to emit a new reservation request, that will not take longer to be effective than using a "fallback" mechanism.

Yes, in the end this boils down to how easy to use the API should be, how many standard use-cases should already be covered by built-in functions, and how much a user of these APIs could do wrong if it isn't already done automatically. This no-fallback method could be provided in addition to a method implementing automatic fallback.

> You are right, and I forgot to include the allocated bandwidth in the segment reservation look, i.e. `ReservationLooks` should contain much more information, like min/max/alloc bandwidth.

Ok if the max and min bandwidth can already be known ahead of time, and it is 100% accurate (probably not), it would resolve the above discussion about performing a fallback when the requested BW is not allowed.

Reply:
> Yes, in the end this boils down to how easy to use the API should be, how many standard use-cases should already be covered by built-in functions, and how much a user of these APIs could do wrong if they are trying to do something that isn't already done automatically. This no-fallback method could be provided in addition to a method implementing automatic fallback.

Not only that: the failure may alter the way that the reservations were originally sorted (e.g. if no bandwidth is avail in AS X, avoid it). This means that the fallback mechanism should be dynamic (evaluated at that point in time, with the error information), which at the moment is not possible unless running a function.

> Ok if the max and min bandwidth can already be known ahead of time, and it is 100% accurate (probably not), it would resolve the above discussion about performing a fallback when the requested BW is not allowed.

min, max, and alloc bandwidth are parameters of the segment reservation. They are accurate for them, but they do not represent the current load of the segments.

I do not agree with any statement that would summarize in "a mechanism in which the client falls back to the next best reservation is good and necessary"; let me explain why, and try to prove it:
- any fallback mechanism in the client is no more efficient than the evaluation of a function dynamically if a reservation setup or renewal fails. If a fallback mechanism could resolve a reservation on a second or even a third try, so would do a function, just by returning this preferences everytime is called.
- Any such preferences expressed in a list that would be present in a fallback mechanism can be represented by the repeated invocation of a function. Thus the fallback mechanism is not necessary.
- Because the last two statements, any behavior of the fallback mechanism is a subset of the error function mechanism.
- Furthermore, the fallback mechanism cannot express dynamic behavior, such as acting on failure of a specific AS (e.g. avoid AS X when possible). This is a very common scenario, and any "good" mechanism should cover it. The fallback mechanism is then not "good".
- As a corollary to the above statement: the fallback mechanism doesn't cover all the behavior that the error function does, and is thus incomplete in the possible behavior cases.